### PR TITLE
Filter push down

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Part 2/3

### CODE REVIEW

File: src/Interpreters/ActionsDAG.cpp [hunk 1]
```diff
@@ -1203,4 +1212,339 @@ ActionsDAG::SplitResult ActionsDAG::splitActionsForFilter(const std::string & co
...
+                    stack.push({.node = child});
+                    break;
+                }
+
+                ++cur.next_child_to_visit;
+            }
+
+            if (cur.next_child_to_visit == cur.node->children.size())
+            {
+                auto & node = actions->nodes.emplace_back(*cur.node);
```
Comment (alesapin) [inline src/Interpreters/ActionsDAG.cpp:1383]: It's a list, so reallocation is not possible, so the pointer is always valid. Maybe add a comment.

File: src/Interpreters/ActionsDAG.cpp [hunk 2]
```diff
@@ -1203,4 +1212,339 @@ ActionsDAG::SplitResult ActionsDAG::splitActionsForFilter(const std::string & co
...
+            }
+        }
+        else
+        {
+            /// Replace predicate result to constant 1.
+            Node node;
+            node.type = ActionType::COLUMN;
+            node.result_name = std::move(predicate->result_name);
+            node.result_type = std::move(predicate->result_type);
+            node.column = node.result_type->createColumnConst(0, 1);
```
Comment (alesapin) [inline src/Interpreters/ActionsDAG.cpp:1482]: Not clear to me :(

File: src/Processors/QueryPlan/Optimizations/optimizeTree.cpp [hunk 3]
```diff
@@ -54,8 +57,13 @@ void optimizeTree(QueryPlan::Node & root, QueryPlan::Nodes & nodes)
             if (!optimization.apply)
                 continue;
 
+            if (max_optimizations_to_apply && max_optimizations_to_apply < total_applied_optimizations)
```
Comment (alesapin) [inline src/Processors/QueryPlan/Optimizations/optimizeTree.cpp:60]: What is it?)

File: src/Processors/QueryPlan/Optimizations/Optimizations.h [hunk 4]
```diff
@@ -38,14 +38,19 @@ size_t trySplitFilter(QueryPlan::Node * node, QueryPlan::Nodes & nodes);
 /// Replace chain `FilterStep -> ExpressionStep` to single FilterStep
 size_t tryMergeExpressions(QueryPlan::Node * parent_node, QueryPlan::Nodes &);
 
+/// Move FilterStep down if possible.
+/// May split FilterStep and push down only part of it.
+size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes);
```
Comment (alesapin) [inline src/Processors/QueryPlan/Optimizations/Optimizations.h:43]: What is returned?

File: src/Interpreters/ActionsDAG.cpp [hunk 5]
```diff
@@ -1203,4 +1212,339 @@ ActionsDAG::SplitResult ActionsDAG::splitActionsForFilter(const std::string & co
...
+            {
+                if (rejected.insert(cur.node).second)
+                    conjunction.rejected.push_back(cur.node);
+            }
+
+            stack.pop();
+        }
+    }
+
+    if (conjunction.allowed.empty())
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Interpreters/ActionsDAG.cpp:1300]: Yes, it is trivial case, when condition does not contain `AND`.
In this case traverse just does not add root node, and it is done manually.

File: src/Interpreters/ActionsDAG.cpp [hunk 6]
```diff
@@ -1203,4 +1212,339 @@ ActionsDAG::SplitResult ActionsDAG::splitActionsForFilter(const std::string & co
...
+    for (const auto * predicate : conjunction)
+    {
+        if (nodes_mapping.count(predicate))
+            continue;
+
+        stack.push({.node = predicate});
+        while (!stack.empty())
+        {
+            auto & cur = stack.top();
+            /// At first, visit all children.
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Interpreters/ActionsDAG.cpp:1367]: Well, the point is that I write DFS using non-recursive way (cause I used to do so  - call stack is expensive).
In classical recursive way, there should be a loop over children:
```c++
DFS(node)
{
    for (auto & child : node->children)
        DFS(child);

    /// do something else
}
```

But in non-recursive implementation we replace call stack to `stack` and recursive `DFS` call to `stack.push`

File: src/Interpreters/ActionsDAG.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Interpreters/ActionsDAG.cpp:1383]: That's a general idea of `ActionsDAG` implementation :)

File: src/Processors/QueryPlan/Optimizations/Optimizations.h (hunk 4 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/Optimizations.h:43]: `the number of updated layers of subtree if some change happened`
It is written in the comment above :)

File: src/Processors/QueryPlan/Optimizations/optimizeTree.cpp (hunk 3 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeTree.cpp:60]: I want to have some protection from having infinite loop of optimizations.
Probably I should set `total_applied_optimizations` to something big like 10000 and throw exception if it was achieved.

**Review by alesapin — APPROVED** [review]

**Pull request merged**